webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/SegmentIndexFileService.scala

```diff
@@ -0,0 +1,181 @@
+import com.scalableminds.webknossos.datastore.storage.{CachedHdf5File, Hdf5FileCache}
+import net.liftweb.common.{Box, Full}
+import play.api.i18n.MessagesProvider
+
+import java.nio.file.{Path, Paths}
+import javax.inject.Inject
+import scala.concurrent.ExecutionContext
+
+class SegmentIndexFileService @Inject()(config: DataStoreConfig,
+                                        binaryDataServiceHolder: BinaryDataServiceHolder,
+                                        dataSourceRepository: DataSourceRepository)(implicit ec: ExecutionContext)
+    extends FoxImplicits
+    with Hdf5Utils
+    with SegmentStatistics {
+  private val dataBaseDir = Paths.get(config.Datastore.baseFolder)
+  private val segmentIndexDir = "segment-index"
+  private val segmentIndexFileExtension = "hdf5"
+
+  private lazy val meshFileCache = new Hdf5FileCache(10)
+
+  def getSegmentIndexFile(organizationName: String, datasetName: String, dataLayerName: String): Box[Path] =
+    for {
+      _ <- Full("")
+      layerDir = dataBaseDir.resolve(organizationName).resolve(datasetName).resolve(dataLayerName)
+      segmentIndexDir = layerDir.resolve(this.segmentIndexDir)
+      files <- PathUtils.listFiles(segmentIndexDir,
+                                   silent = false,
+                                   PathUtils.fileExtensionFilter(segmentIndexFileExtension))
+      file <- files.headOption
+    } yield file
+
+  def readSegmentIndex(organizationName: String,
+                       datasetName: String,
+                       dataLayerName: String,
+                       segmentId: Long): Fox[(Array[Vec3Int], Vec3Int)] =
+    for {
+      segmentIndexPath <- getSegmentIndexFile(organizationName, datasetName, dataLayerName).toFox
+      segmentIndex = meshFileCache.withCache(segmentIndexPath)(CachedHdf5File.fromPath)
+      hashFunction = getHashFunction(segmentIndex.reader.string().getAttr("/", "hash_function"))
+      nBuckets = segmentIndex.reader.uint64().getAttr("/", "n_hash_buckets")
+      mag <- Vec3Int.fromArray(segmentIndex.reader.uint64().getArrayAttr("/", "mag").map(_.toInt)).toFox
+      bucketIndex = hashFunction(segmentId) % nBuckets
+      bucketOffsets = segmentIndex.reader.uint64().readArrayBlockWithOffset("hash_bucket_offsets", 2, bucketIndex)
+      bucketStart = bucketOffsets(0)
+      bucketEnd = bucketOffsets(1)
+      segmentExists = bucketEnd - bucketStart != 0
+      topLefts <- Fox.runIf(segmentExists)(for {
+        _ <- Fox.successful(())
+        buckets = segmentIndex.reader
+          .uint64()
+          .readMatrixBlockWithOffset("hash_buckets", (bucketEnd - bucketStart + 1).toInt, 3, bucketStart, 0)
+        bucketLocalOffset = buckets.map(_(0)).indexOf(segmentId)
+        _ <- bool2Fox(bucketLocalOffset >= 0)
+        topLeftStart = buckets(bucketLocalOffset)(1)
+        topLeftEnd = buckets(bucketLocalOffset)(2)
+        topLefts = segmentIndex.reader
+          .uint16() // Read datatype from attributes?
```

Comment:
@markbader ping

Reply:
Ok we found out that it is indeed the metadata field `dtype_bucket_entries`. @frcroth please add an assertion that the value of that metadata field is indeed uint16 so that we can get a readable error message otherwise, and not undefined behavior by reading the wrong bytes.

Reply:
I changed my mind, this can be a quick follow-up for later https://github.com/scalableminds/webknossos/issues/7666 as there are currently no non-uint16 files and we should merge this PR ASAP. I’ll do some more testing all around tomorrow.